### PR TITLE
Bit-exact Nearest Neighbor Resizing

### PR DESCRIPTION
**Pull request by Yosshi999**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by terfendail** [comment]
@Yosshi999 Could you please check the `Nearest8U_vsNonExact` failure reason on Win64? That could be near-edge issue we discussed

**Comment by Yosshi999** [comment]
I thought x2 or x0.5 resizing is bit-exact even if the method is non-exact version because multiplication by 0.5 or 2 will not modify fraction fields of floating point data. But it seems I have a misunderstanding.

File: modules/imgproc/src/resize.cpp [hunk 1]
```diff
@@ -1207,10 +1208,15 @@ static void resizeNN_bitexact( const Mat& src, Mat& dst, double _fx, double _fy
     area.allocate(x_ofse, dsize.width, CV_SIMD_WIDTH);
     area.commit();
 
+    std::cout << "ifx: " << std::hex << ifx.raw() << std::endl;
     for( int x = 0; x < dsize.width; x++ )
     {
         int sx = int32_t( ((ifx * static_cast<uint32_t>(x)) >> 32) << 32 ); // equivalent to cvFloor
```
Comment (Yosshi999) [inline modules/imgproc/src/resize.cpp:1214]: It seems `(x >> 32) << 32` is optimized to `x` in the windows build.

File: modules/imgproc/test/test_resize_bitexact.cpp [hunk 2]
```diff
@@ -152,4 +152,33 @@ TEST(Resize_Bitexact, Linear8U)
...
+
+    Mat src_color = imread(cvtest::findDataFile("shared/lena.png"));
+    Mat src_gray; cv::cvtColor(src_color, src_gray, COLOR_BGR2GRAY);
+
+    int indices = sizeof(depths)/sizeof(depths[0]);
+    double normINF;
+    for (int i = 0; i < indices; i++)
+    {
+        Mat mat_color, mat_gray;
+        src_color.assignTo(mat_color, depths[i]);
```
Comment (alalek) [inline modules/imgproc/test/test_resize_bitexact.cpp:174]: > assignTo

This method is "internal" and should not be used.
Use `.convertTo()` instead.

File: modules/imgproc/test/test_resize_bitexact.cpp [hunk 3]
```diff
@@ -152,4 +152,33 @@ TEST(Resize_Bitexact, Linear8U)
...
+static double checkNearest8U_vsNonExact(const Mat& src)
+{
+    Mat dstExact; cv::resize(src, dstExact, Size(), 2, 1, INTER_NEAREST_EXACT);
+    Mat dstNonExact; cv::resize(src, dstNonExact, Size(), 2, 1, INTER_NEAREST);
+
+    return cv::norm(dstExact, dstNonExact, NORM_INF);
+}
+TEST(Resize_Bitexact, Nearest8U_vsNonExact)
+{
+    int depths[] = {CV_8U, CV_16U, CV_32F, CV_64F};
```
Comment (alalek) [inline modules/imgproc/test/test_resize_bitexact.cpp:164]: Please make this as test parameter:
- use `MatDepth` wrapped test type
- look for `INSTANTIATE_TEST_CASE_P`

File: modules/imgproc/src/resize.cpp [hunk 4]
```diff
@@ -1104,6 +1105,117 @@ resizeNN( const Mat& src, Mat& dst, double fx, double fy )
...
+    ufixedpoint64 ify(softdouble::one() / softdouble(_fy));
+    Size ssize = src.size(), dsize = dst.size();
+    cv::utils::BufferArea area;
+    int* x_ofse = 0;
+    area.allocate(x_ofse, dsize.width, CV_SIMD_WIDTH);
+    area.commit();
+
+    for( int x = 0; x < dsize.width; x++ )
+    {
+        int sx = (ifx * static_cast<uint32_t>(x)).raw() >> 32; // equivalent to cvFloor
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp:1212]: May be we should implement `ufixedpoint64::cvFloor() const` method instead?

File: modules/imgproc/perf/perf_resize.cpp [hunk 5]
```diff
@@ -254,4 +254,30 @@ PERF_TEST_P(MatInfo_Size_Scale_NN, ResizeNN,
     SANITY_CHECK_NOTHING();
 }
 
+PERF_TEST_P(MatInfo_Size_Scale_NN, ResizeNNExact,
+    testing::Combine(
+        testing::Values(CV_8UC1, CV_8UC2, CV_8UC4),
+        testing::Values(szVGA, szqHD, sz720p, sz1080p, sz2160p),
+        testing::Values(2.4, 3.4, 1.3)
```
Comment (Yosshi999) [inline modules/imgproc/perf/perf_resize.cpp:261]:
```suggestion
    testing::Combine(
        testing::Values(CV_8UC1, CV_8UC3, CV_8UC4),
        testing::Values(sz720p, sz1080p),
        testing::Values(0.25, 0.5, 2.0)
```

**Comment by vpisarev** [comment]
@Yosshi999, thank you for the contribution!
Can you, please, implement the following algorithm instead https://github.com/opencv/opencv/pull/17068#issuecomment-657136140?

**Comment by alalek** [comment]
> Can you, please, implement the following algorithm instead

This is completely different algorithm.

Bit-exact implementation should follow this rule and must provide minimal possible difference between:
- bit-exact result of 8U input
- convert to double/floats, run original version on floating-point data, convert result back to 8U.

Without changing of the base NN implementation this should not be called bit-exact.

**Comment by Yosshi999** [comment]
According to the discussion in https://github.com/opencv/opencv/pull/17068, it seems that there are some options:
* Define it as a new method like `INTER_NEAREST_PILLOW` or `INTER_NEAREST_CENTER_PIX_CONVENTION` or something. (@catree said we should avoid using the name of a library so I prefer `INTER_NEAREST_CENTER_...` if this option is chosen)
* Put it into the current implementation of `INTER_NEAREST` (@vpisarev wants to do this, but @alalek disagrees?)

In these cases above, I will remove the suffix `_EXACT` and related tests even though this is intended to be bit-exact. I think we don't have to create non-exact version because this doesn't use softfloat so the overhead is not large.

Or:
* Revert to ee74e10 and define it as `INTER_NEAREST_EXACT`

I will vote for the second one.

**Review by terfendail — APPROVED** [review]
Implementation looks fine for me. The only thing to discuss is the name of the method enum. However it is outside of this PR intention scope.

**Pull request merged**